### ISSUE TITLE
ClojureScript compilation support

### ISSUE THREAD

**Issue by ajoberstar** [gradle-clojure/gradle-clojure#28]

**Comment by pbzdyl** [comment]
@gradle-clojure/developers What would be the best way to provide ClojureScript compiler options? The compiler supports a lot of different options and having them supported as the Gradle CLJS compiler properties would make the configuration quite involved (including properties/entries that can have values of different types, e.g. `source-map` might be either boolean or a string path).

I can see following options:
- external edn file with the compiler options (possible issue is having dynamic variables interpolated in the configuration)?
- a string with edn? (cumbersome to escape e.g. string quotes but allows for dynamic variables via `GString`)
- mirror all compiler options as ClojureScript compile task properties

**Comment by ajoberstar** [comment]
I think we'd probably have to mirror the options on the task. Not too familiar with the CLJS compiler, so maybe that gets us into too much trouble though.

**Comment by pbzdyl** [comment]
@ajoberstar I am working on a prototype of basic ClojureScript compile task. I have a few questions regarding the current code design and how ClojureScript support should be added. Would you have a few minutes to answer them?

1. Currently there is a module for the Clojure Gradle plugin. Shall we have ClojureScript tasks in this module or create another one?

2. Clojure source set defaults are configured in `ClojureBasePlugin`. Should ClojureScript defaults be configured there too or should that code be moved to `ClojurePlugin` and `ClojureScriptPlugin` respectively? Correct me if I'm wrong but with both configured in `ClojureBasePlugin` they will be both always configured even if a user applies only one of the plugins.

If you think it would be better to have an online chat, please, do let me know when and where it would be best to talk.

Thank you in advance,
Piotr

**Comment by ajoberstar** [comment]
> 1. Currently there is a module for the Clojure Gradle plugin. Shall we have ClojureScript tasks in this module or create another one?

Assuming we won't need to add a lot of additional dependencies, I'd lean towards keeping one module. Depending on how much code this adds, we might want to break up the packages differently, but no need to do that pre-emptively.

> 2. Clojure source set defaults are configured in ClojureBasePlugin. Should ClojureScript defaults be configured there too or should that code be moved to ClojurePlugin and ClojureScriptPlugin respectively? Correct me if I'm wrong but with both configured in ClojureBasePlugin they will be both always configured even if a user applies only one of the plugins.

I think we'll want a `ClojureScriptBasePlugin` and a `ClojureScriptPlugin` to mirror the Clojure setup. The idea would be that if you just want the capability to use a language, you can apply the base, but generally you'll apply the "main" one to get the conventional build setup.

One of my struggles on ClojureScript is what the minimal version of that looks like. Do you need some Clojure support to use ClojureScript? Or is it only if you need macros that you would have to write Clojure code for an otherwise ClojureScript-only project?

There's also going to be the question of if/how a project gets packaged by default. For ClojureScript-only would you just produce the compiled output? And for Clojure+ClojureScript also package it into a JAR?

Personally, I'll need to do some more review of how the CLJS support works in Leiningen and Boot once I get through the 0.3.0 stuff.

I might be getting ahead of myself with some of those questions, but either way I can do an online chat if you're ready to discuss more. For me 7pm-10pm (CDT) would be easiest during the week, otherwise I can be more available on weekends.  Hangouts and Skype are a couple options.

**Comment by pbzdyl** [comment]
Thank you @ajoberstar for your reply. Unfortunately, I didn't have as much time last week as I planned so I didn't make too much progress. I have pushed a very rough version in pbzdyl/cljs branch. It supports most of the compiler options. There are still a lot of issues to be solved, for example:
- a classpath issue with Guava - I had to explicitly add Guava in the plugin compile path, otherwise an older version is picked up causing exceptions. So far I was unable to find where older Guava comes from, `gradlew dependencies` doesn't show it when I don't add it
- DSL for adding multiple configuration options like foreign lib or modules.

I have added a single test showing the simplest use case of the plugin.

**Comment by ajoberstar** [comment]
Cool, I should be able to take a look sometime this week.

**Comment by ajoberstar** [comment]
Was going to take a look through this today, but ended up wrapping up the 0.3.0 issues and release. Hope to look this week.

**Comment by pbzdyl** [comment]
No worries, I have been busy with other things the whole week. Hopefully I will make more progress this week.

**Comment by ajoberstar** [comment]
Took a read through, and this is really good progress. Getting all of those compiler options put together is quite the ordeal.

On the Guava issue, where do you see that come up? I'm not seeing any Guava references in your code. I expect there's probably a version of Guava in Gradle itself somewhere, but not positive what version.

For the DSL of multi-option things like foreign libs, you might be able to use a [NamedDomainObjectContainer](https://docs.gradle.org/current/javadoc/org/gradle/api/NamedDomainObjectContainer.html). It might be overkill in this scenario, but would also probably handle the DSL needs. You can get one from the [Project#container(Class)](https://docs.gradle.org/current/javadoc/org/gradle/api/Project.html#container(java.lang.Class)) method.

Once I get a little time using 0.3.0 under my belt, I think I'll be reviewing leing-cljsbuild and figwheel to get a better sense of what the expected overall ClojureScript environment will need to involve.

Let me know if there's anything else you could use some help on in the meantime.

**Comment by pbzdyl** [comment]
Thank you @ajoberstar for taking the time to review my code.

I will try to use NamedDomainObjectContainer but first I am trying to fix the Guava classpath issue.

What I have found is that Gradle repackages Guava under it's own package (`com.google.common.base` -> `org.gradle.internal.impldep.com.google.common.base`). Unfortunately, it looks that the Guava jar is also bundled in Gradle bundle used by Gradle wrapper:

`~/.gradle/wrapper/dists/gradle-4.3-rc-1-bin/4dpv3r3e5mwtficz0bbnrpkn4/gradle-4.3-rc-1/lib/guava-jdk5-17.0.jar`

I tried to simply delete it but then Gradle fails:

```
java.lang.IllegalArgumentException: Cannot find JAR 'guava-jdk5-17.0.jar' required by module 'gradle-base-services' using classpath or distribution directory '/Users/Badyl/.gradle/wrapper/dists/gradle-4.3-rc-1-bin/4dpv3r3e5mwtficz0bbnrpkn4/gradle-4.3-rc-1'
	at org.gradle.api.internal.classpath.DefaultModuleRegistry.findDependencyJar(DefaultModuleRegistry.java:262)
	at org.gradle.api.internal.classpath.DefaultModuleRegistry.findDependencyJars(DefaultModuleRegistry.java:152)
	at org.gradle.api.internal.classpath.DefaultModuleRegistry.module(DefaultModuleRegistry.java:142)
	at org.gradle.api.internal.classpath.DefaultModuleRegistry.loadOptionalModule(DefaultModuleRegistry.java:122)
	at org.gradle.api.internal.classpath.DefaultModuleRegistry.loadModule(DefaultModuleRegistry.java:106)
	at org.gradle.api.internal.classpath.DefaultModuleRegistry.getModule(DefaultModuleRegistry.java:86)
	at org.gradle.api.internal.classpath.DefaultModuleRegistry.getModules(DefaultModuleRegistry.java:160)
	at org.gradle.api.internal.classpath.DefaultModuleRegistry.access$100(DefaultModuleRegistry.java:45)
	at org.gradle.api.internal.classpath.DefaultModuleRegistry$DefaultModule.getRequiredModules(DefaultModuleRegistry.java:296)
	at org.gradle.api.internal.classpath.DefaultModuleRegistry$DefaultModule.collectRequiredModules(DefaultModuleRegistry.java:321)
	at org.gradle.api.internal.classpath.DefaultModuleRegistry$DefaultModule.getAllRequiredModules(DefaultModuleRegistry.java:313)
	at org.gradle.api.internal.DefaultClassPathProvider.findClassPath(DefaultClassPathProvider.java:32)
	at org.gradle.api.internal.DefaultClassPathRegistry.getClassPath(DefaultClassPathRegistry.java:34)
	at org.gradle.launcher.bootstrap.ProcessBootstrap.runNoExit(ProcessBootstrap.java:49)
	at org.gradle.launcher.bootstrap.ProcessBootstrap.run(ProcessBootstrap.java:37)
	at org.gradle.launcher.GradleMain.main(GradleMain.java:23)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.wrapper.BootstrapMainStarter.start(BootstrapMainStarter.java:31)
	at org.gradle.wrapper.WrapperExecutor.execute(WrapperExecutor.java:108)
	at org.gradle.wrapper.GradleWrapperMain.main(GradleWrapperMain.java:61)
```

Thus I guess it is required by Gradle even though it also contains repackaged version.

I am not sure how I should handle it. The options I see are:
- don't use ClojureWorker but run ClojureScript compilation in a new JVM process (so we have full control of its classpath)
- use parent-last policy for shimdandy classpath - I am not even sure it's possible out of the box in shimdandy

I am also not sure if having Guava bundled in Gradle is intended behaviour and maybe `gradle-base-services` should use the repackaged classes instead and drop dependency on standard Guava.

I will post updates when I make any progress on this.

**Comment by ajoberstar** [comment]
OK, so you're seeing the classpath issue when you run the ClojureScript compiler in a `ClojureWorker`?

One of the complications to dropping the worker is that the inputs to the compiler get more complicated. With the worker we can pass that CompilerOptions object all the way into our Clojure code and fairly easily format it into the EDN map that is needed in the end.

Some kind of parent-last or blacklisting classloader might work. I was using one in a different iteration of the worker changes. I think I blew away the commits though.

**Comment by pbzdyl** [comment]
[The response from one of Gradle developers](https://discuss.gradle.org/t/guava-in-gradles-classpath/24861/4) makes me think that manipulating the classpath (either directly from `ClojureWorker` or somewhere in shimdandy) might clash with the classpath filtering mechanism in Gradle.

For now I will go with calling external process instead of using workers. I will dump all the function call parameters into a temporary EDN file and read it inside `clojurescript_compiler.clj` to get all the parameters instead of going through command parameters.

**Comment by ajoberstar** [comment]
Yeah, that makes sense. I'll keep looking into the classpath issues since I've noticed some problems in the REPL too.

**Comment by pbzdyl** [comment]
I have pushed updated version of the basic ClojureScript compile task.

The syntax of the compile options can be seen in [the test case](https://github.com/gradle-clojure/gradle-clojure/blob/pbzdyl/cljs/modules/gradle-clojure-plugin/src/compatTest/projects/BasicClojurescriptProjectTest/build.gradle#L30-L50).

I have also switched from ClojureWorker to calling Java process and the Guava classpath issue is gone.

Any feedback is welcomed! :)

**Comment by ajoberstar** [comment]
Plan to look at this over the weekend.

**Comment by ajoberstar** [comment]
Only took a brief look (since it's hard to tell what changed from the last version). If we can't get the workers to operate cleanly, I think this is a good option. The only way to cut the code down would be to do object serialization like the worker (but safer, only sharing the specific classes we want to serialize). Not worth looking into now. I'm going to do some digging on the workers to see if I can clean up some of the behavior.

**Comment by ajoberstar** [comment]
If you still have a copy of your worker-based version with the Guava issue (maybe in your reflog), I'd appreciate if you could push it to a separate branch. I'm curious if something like #43 is causing some issues here.

**Comment by ajoberstar** [comment]
I put together a test case for your guava issue. And will post it to the Gradle forum on the thread you opened.

**Comment by pbzdyl** [comment]
Thank you @ajoberstar. I have pushed my old version of the plugin to `pbzdyl/cljs-clojureworker-issue` branch.

Does the test case for guava issue contain all the necessary files? I think there might be a file missing in `buildSrc` directory.

**Comment by pbzdyl** [comment]
Regarding your suggestion to use serialisation I am not sure if it will cut too much of the code as even when I was passing the objects directly with ClojureWorker I had to convert the objects into appropriate Clojure data structures and [it required quite a lot of code too](https://github.com/gradle-clojure/gradle-clojure/blob/pbzdyl/cljs-clojureworker-issue/modules/gradle-clojure-tools/src/main/clojure/gradle_clojure/tools/clojurescript_compiler.clj#L6-L93). And with serialisation I would have to add our Clojure plugin and Gradle jars to the classpath of the CLJS compiler process. Dumping ("serialising") all the options to EDN decouples the external process which I think is good.

**Comment by ajoberstar** [comment]
> Does the test case for guava issue contain all the necessary files? I think there might be a file missing in buildSrc directory.

Good catch. My gitignore was misconfigured, should be there now.

> Regarding your suggestion to use serialisation I am not sure if it will cut too much of the code as even when I was passing the objects directly with ClojureWorker I had to convert the objects into appropriate Clojure data structures and it required quite a lot of code too. And with serialisation I would have to add our Clojure plugin and Gradle jars to the classpath of the CLJS compiler process. Dumping ("serialising") all the options to EDN decouples the external process which I think is good.

Oh, yeah, I forgot how involved that still was before.

**Comment by ajoberstar** [comment]
FYI, I do have the Guava issue fixed in #54. I still hope Gradle will do some improvements to the worker isolation, but I think we can get around it.

**Comment by ajoberstar** [comment]
I'm finally starting on this now. Just got your branch updated with the code from master. That was just to the point of getting it to build, but want to make some updates to consolidate with changes made in master since. I'll probably be moving my `ClojureCompileOptions` -> EDN stuff to use the approach you used in `CljsEdnUtils`. Seems like a better way to keep that centralized.

**Comment by ajoberstar** [comment]
Started just browsing through the compile options documentation and looking at what you put together. Wowzers... That's a lot of options... 

I definitely think there's some opportunity to make this easier for buildscript authors to use by putting some Gradle modelling over the top. But I think we should do a few things before we go down that road:

1. Talk to the Gradle folks about the "variant" support they have, which seems like it could help with the idea of modelling multiple builds.
1. Do more research on the CLJS compiler options, lein-figwheel, boot-cljs, and shadow-cljs. Look at how they all approach this topic.
1. Get some feedback from the Clojure community on what common patterns for CLJS builds there are. 
1. Modelling would target making it easy to do common things, without having to configure all of the compiler options yourself.

As for 0.4.0, I think what I'd like to do is release it with a `gradle-clojure.clojurescript-base` plugin and **not** provide a `gradle.clojure-clojurescript` plugin. Base would just provide:

1. Essentially what Piotrek already added. A clojurescriptCompile task for each source set. Let people configure the options themselves.
1. (After I do some research to see how easy this would be) minimal Figwheel support. Just allowing them to provide the task and configure it themselves. That way this would provide a usable development environment, but leave some config to the user.

**Comment by ajoberstar** [comment]
Fixed by #82.
